crates/store/src/state/state_db.rs

```diff
@@ -94,152 +274,306 @@ impl<'a> StateTree<'a> {
         Ok(())
     }
 
-    pub fn get_merkle_state(&self) -> AccountMerkleState {
-        AccountMerkleState::new_builder()
-            .merkle_root(self.tree.root().pack())
-            .count(self.account_count.pack())
-            .build()
-    }
-
-    /// Detach block state from state tree
-    pub fn detach_block_state(&mut self) -> Result<()> {
-        let block_number = match self.context {
-            StateContext::DetachBlock(block_number) => block_number,
-            ctx => return Err(anyhow!("Wrong context in detach block state: {:?}", ctx)),
-        };
-
-        // reset states to previous value
-        let parent_block_number = block_number.saturating_sub(1);
-        let reverted_key_values: Vec<_> = self
-            .db()
-            .iter_block_state_record(block_number)
-            .map(|record_key| {
-                let state_key = record_key.state_key();
-                let last_value = self
-                    .db()
-                    .get_history_state(parent_block_number, &state_key)
-                    .unwrap_or_else(H256::zero);
-                (state_key, last_value)
-            })
-            .collect();
-        for (state_key, last_value) in reverted_key_values {
-            self.update_raw(state_key, last_value)?;
-        }
-
-        // remove block's state record
-        self.db().remove_block_state_record(block_number)?;
+    /// Return last state root
+    /// Invoke `calculate_root` will update this value
+    pub fn last_state_root(&self) -> H256 {
+        self.last_state_root
+    }
+
+    /// create snapshot
+    pub fn snapshot(&mut self) -> usize {
+        let id = self.next_revision_id;
+        self.next_revision_id += 1;
+        self.revisions.push_back(Revision {
+            id,
+            journal_len: self.journal.len(),
+        });
+        id
+    }
+
+    /// revert to a snapshot
+    pub fn revert(&mut self, id: usize) -> Result<()> {
+        // find revision
+        let rev_index = self
+            .revisions
+            .binary_search_by_key(&id, |r| r.id)
+            .map_err(|_id| anyhow!("Invalid revision id"))?;
+        let rev = &self.revisions[rev_index];
+
+        // replay to revert journal
+        let revert_entries = self.journal.revert_entries(rev.journal_len);
+        for entry in revert_entries.into_iter().rev() {
+            entry.revert(self);
```

Comment:
Why not implement snapshot/revert by simply cloning/restoring all the dirty data fields? `im::HashMap` is supposed to be cheap to clone. It would be easier to implement and might be faster too.

Reply:
Yeah, the performance should be samilar. I just pick one between the two choices.